jcl/src/openj9.jvm/share/classes/com/ibm/jvm/CRIUSupport.java

```diff
@@ -23,17 +23,51 @@
  *******************************************************************************/
 package com.ibm.jvm;
 
+import java.nio.file.Paths;
+
 /**
  * CRIU Support API
  */
 public final class CRIUSupport {
 	
-	private static final boolean criuSupportEnabled = isCRIUSupportEnabledImpl();
+	/**
+	 * CRIU operation return code
+	 */
+	public enum CRIUOpRetCode {
+		/**
+		 * The operation was successful.
+		 */
+		SUCCESS,
+		/**
+		 * The requested operation is unsupported.
+		 */
+		UNSUPPORTED_OPERATION,
+		/**
+		 * The arguements provided for the operation are invalid.
+		 */
+		INVALID_ARGUEMENTS,
+		/**
+		 * A system failure occured when attempting to checkpoint the JVM.
+		 */
+		SYSTEM_CHECKPOINT_FAILURE,
+		/**
+		 * A JVM failure occured when attempting to checkpoint the JVM.
+		 */
+		JVM_CHECKPOINT_FAILURE,
+		/**
+		 * A non-fatal JVM failire occured when attempting to restore the JVM.
+		 */
+		JVM_RESTORE_FAILURE,
```

Comment:
Do `CHECKPOINT_FAILURE`/`RESTORE_FAILURE` include the hooks failing?  How will a hook throwing an exception be handled in this api?

Reply:
I think the model will need to grow as there will be user level hooks as well, right?  Or are they expected to be handled prior to calling into `checkPointJVM()`?

Reply:
> Do CHECKPOINT_FAILURE/RESTORE_FAILURE include the hooks failing?

The `JVM_CHECKPOINT_FAILURE` and `JVM_RESTORE_FAILURE` iare used to indicate a failure in JVM hooks or other JVM failures (nativeOOM, etc.) 

>  How will a hook throwing an exception be handled in this api?

The intention is that we would catch those exceptions and return `JVM_CHECKPOINT_FAILURE` and `JVM_RESTORE_FAILURE`.

> I think the model will need to grow as there will be user level hooks as well, right? Or are they expected to be handled prior to calling into checkPointJVM()?

The current model is that the application is responsible for their own hooks. So we have a stack model that looks something like this:

On checkpoint:
1: Application_checkpoint_hooks
2: JVM_checkpoint_hooks
3: criu_dump()

On restore
1: criu_dump()
2: JVM_restore_hooks
3. Application_restore_hooks

Reply:
Ok, so we're explicitly not getting involved in `Application_restore_hooks`.  

For failure cases, we'll want more info to tell why a failure occurred.  Tracepoints will work for us but being able to give more info to the user of the API in the case of `FAILURE` would be a big help with them understanding why a checkpoint/restore operation didn't succeed.

Assuming this is an initial sketch for building experience with the API, enhancing the error messages can be handled under a later PR

Reply:
> Assuming this is an initial sketch for building experience with the API, enhancing the error messages can be handled under a later PR

Yes this the intention. We are not still fully certain what the user expects in these cases. We know that at the very least the user needs to know if a failure occurred creating a checkpoint or restoring a checkpoint (and if its successful)